### PR TITLE
libobs: Update macOS platform implementation

### PR DESCRIPTION
**Pull request by PatTheMav**
### Description
Cleans up the source code by sorting functions by topic and also removes macros for hotkey implementation, removing the need to disable compiler warnings.

Key codes and descriptions are now contained in arrays which either use the OBS key code or the macOS key code as lookup keys with all necessary context information directly associated with the key.

The implementation also uses the zero-initialization rule for arrays of structs in such a way that the struct's "is_valid" boolean is "false" by default so that only values explicitly initialized with a key are ever "valid".

### Motivation and Context
Make code more readable and easier to maintain by removing use of macros and also sorting functions into logical units (source code `MARK`s also make it easier to navigate through code in Xcode).

### How Has This Been Tested?
Tested on macOS 14.

### Types of changes
- Code cleanup (non-breaking change which makes code smaller or more readable)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by RytoEX** [comment]
Would be helpful to have review from a macOS dev. @gxalpha @jcm93

File: libobs/obs-cocoa.m
Comment (jcm93) [inline libobs/obs-cocoa.m:382]: Is there a reason this array has size 255?

Comment (PatTheMav) [reply libobs/obs-cocoa.m:382]: Probably an oversight, using `OBS_KEY_LAST_VALUE` should work here as well.

File: libobs/obs-cocoa.m
Comment (jcm93) [inline libobs/obs-cocoa.m:779]: We lost the comment here about 0x104 being related to caps lock, which may have been useful/informative for reading the code. Maybe worth defining 0x104 as a constant?

**Comment by jcm93** [comment]
Can confirm that hotkeys still in good working order.

File: libobs/obs-cocoa.m
Comment (jcm93) [reply libobs/obs-cocoa.m:779]: Per the docs [here](https://developer.apple.com/documentation/coreservices/1390584-uckeytranslate), this is derived from some sort of bit shift from a Carbon modifier key code masked with `0xFF`, but I can't figure out the exact derivation for caps lock in particular...

File: libobs/obs-cocoa.m
Comment (PatTheMav) [reply libobs/obs-cocoa.m:382]: Ah wait, I'm mistaken - that specific array uses the macOS key codes as its key, so there is no single "largest value" (like `OBS_KEY_LAST_VALUE`) that could be used here.

In theory one could use `0x7F` here (as `0x7E` seems to be the highest key code defined in Carbon's `Events.h` (and that header hasn't changed in ages), `0xFF` is ensured to encapsulate all current key codes though.

File: libobs/obs-cocoa.m
Comment (PatTheMav) [reply libobs/obs-cocoa.m:779]: Maybe a stupid question, but how can _any_ value logically combined with `0xFF` ever produce a value larger than `0xFF` (as `0x104` is)?

That value seems invalid to me, because the only value that remains from the bit mask is bit 3 set to `1` (`0x40`).

The `modifiers` member of the `EventRecord` struct is a `UInt16` (aka `EventModifiers`), and caps lock (aka alphaLock) is bit 10 (`0x400`) and if I shift that value by 8 I get `0x4`.

Comment (PatTheMav) [reply libobs/obs-cocoa.m:779]: Funnily enough what might happen here is that the 9th bit might just be ignored by macOS and it "sees" just the `0x04` part, which would align again.

Comment (jcm93) [reply libobs/obs-cocoa.m:779]: I presume that must be the case; it looks like all of the other Carbon modifier key enums would also have ignored bits when shifted (though they do all end up being unique at least); I have no idea why this would be done this way.

It looks like the purpose of this block of code in general is to ensure that the string representation of the chosen hotkey is always capitalized, I guess so OBS doesn't imply that capitalization state matters for letter-key keybinds.

I confirmed that just putting in `4` works to achieve the same effect.

It seems like we definitely wouldn't want to import Carbon just for the sake of having the actual `alphaLock` available for defining this goofy constant, so maybe we just do something like `UInt32 carbonCapsLockBitMask = 0x4` and provide `carbonCapsLockBitMask` to `UCKeyTranslate` for readability?

File: libobs/obs-cocoa.m
Comment (jcm93) [reply libobs/obs-cocoa.m:382]: Yeah I suppose that's fine. I guess a 0x7E minimally sized array would take up a bit less space, but unlikely to matter. Regardless I think either a brief code comment or a named constant for the size might be good here, just to make it easier to understand at a glance.

File: libobs/obs-cocoa.m
Comment (PatTheMav) [reply libobs/obs-cocoa.m:779]: We already include the Carbon header anyway for the other key codes, so we can use the "official" values.

File: libobs/obs-cocoa.m
Comment (PatTheMav) [reply libobs/obs-cocoa.m]: I use the keyCode bit mask now (which is 8 bits wide) and downshift it by 8 bytes to effectively get `0xFF`, but it should make it more obvious (as that implies that a keyCode _has_ to fit into the width of the mask for a Carbon event).

File: libobs/obs-cocoa.m
Comment (PatTheMav) [reply libobs/obs-cocoa.m]: I use `alphaLock` now and downshift + mask it as expected by the documentation for `UCKeyTranslate`.

**Review by jcm93 — COMMENTED** [review]
LGTM after latest push.

**Review by RytoEX — APPROVED** [review]
@jcm93 says this looks good to them, and this is isolated to macOS. If there are unforeseen issues, it will only affect macOS.

**Pull request merged**
